In [1]:
import pandas as pd   #importing libraries
import numpy as np
import os

In [2]:
#importing dataframes
path = r'C:\Instacart Basket Achievement 4 NEW'

In [3]:
path

'C:\\Instacart Basket Achievement 4 NEW'

In [4]:
df_ords_prods_merge = pd.read_pickle(os.path.join(path, '02 Data' , 'Prepared Data', 'orders_products_merged.pkl'))

In [5]:
#creating subset to work with
df = df_ords_prods_merge[:1000000]  #first 1,000,000 rows in the dataframe

In [6]:
df.shape

(1000000, 14)

In [7]:
def price_label(row):   #define a function

  if row['prices'] <= 5:
    return 'Low-range product'
  elif (row['prices'] > 5) and (row['prices'] <= 15):   #elif means else if
    return 'Mid-range product'
  elif row['prices'] > 15:
    return 'High range'
  else: return 'Not enough data'

In [8]:
df['price_range'] = df.apply(price_label, axis=1)  #apply the function

C:\Users\nehma\AppData\Local\Temp\ipykernel_13400\299912984.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['price_range'] = df.apply(price_label, axis=1)  #apply the function


In [9]:
df['price_range'].value_counts(dropna = False)

price_range
Mid-range product    756450
Low-range product    243550
Name: count, dtype: int64

In [10]:
df['prices'].max()  #most expensive item in subset

14.8

In [11]:
df.loc[df['prices'] > 15, 'price_range_loc'] = 'High-range product'  #using loc function because of warning

C:\Users\nehma\AppData\Local\Temp\ipykernel_13400\3122393354.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df.loc[df['prices'] > 15, 'price_range_loc'] = 'High-range product'  #using loc function because of warning
C:\Users\nehma\AppData\Local\Temp\ipykernel_13400\3122393354.py:1: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise in a future error of pandas. Value 'High-range product' has dtype incompatible with float64, please explicitly cast to a compatible dtype first.
  df.loc[df['prices'] > 15, 'price_range_loc'] = 'High-range product'  #using loc function because of warning


In [12]:
df.loc[(df['prices'] <= 15) & (df['prices'] > 5), 'price_range_loc'] = 'Mid-range product' 

In [13]:
df.loc[df['prices'] <= 5, 'price_range_loc'] = 'Low-range product'

In [14]:
df['price_range_loc'].value_counts(dropna = False)

price_range_loc
Mid-range product    756450
Low-range product    243550
Name: count, dtype: int64

In [15]:
#on the entire dataframe
df_ords_prods_merge.loc[df_ords_prods_merge['prices'] > 15, 'price_range_loc'] = 'High-range product'

C:\Users\nehma\AppData\Local\Temp\ipykernel_13400\463402567.py:2: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise in a future error of pandas. Value 'High-range product' has dtype incompatible with float64, please explicitly cast to a compatible dtype first.
  df_ords_prods_merge.loc[df_ords_prods_merge['prices'] > 15, 'price_range_loc'] = 'High-range product'


In [16]:
df_ords_prods_merge.loc[df_ords_prods_merge['prices'] <= 5, 'price_range_loc'] = 'Low-range product'

In [17]:
df_ords_prods_merge['price_range_loc'].value_counts(dropna = False)

price_range_loc
NaN                   21860860
Low-range product     10126321
High-range product      417678
Name: count, dtype: int64

In [18]:
df_ords_prods_merge['orders_day_of_week'].value_counts(dropna = False) 

orders_day_of_week
0    6204182
1    5660230
6    4496490
2    4213830
5    4205791
3    3840534
4    3783802
Name: count, dtype: int64

### Saturday is the busiest day of the week, while Wednesday is the slowest day of week

In [19]:
#creating a new column
result = []

for value in df_ords_prods_merge["orders_day_of_week"]:
  if value == 0:
    result.append("Busiest day")
  elif value == 4:
    result.append("Least busy")
  else:
    result.append("Regularly busy")

In [20]:
result

['Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Regularly busy',
 'Busiest day',
 'Regularly busy',
 'Reg

In [21]:
df_ords_prods_merge['busiest_day'] = result

In [22]:
df_ords_prods_merge['busiest_day'].value_counts(dropna = False)

busiest_day
Regularly busy    22416875
Busiest day        6204182
Least busy         3783802
Name: count, dtype: int64

In [23]:
df_ords_prods_merge.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_prior_order  product_id  add_to_cart_order  reordered _merge  \
0                     NaN         196                  1          0   both   
1                    15.0         196                  1          1   both   
2                    21.0         196                  1          1   both   
3                    29.0         196                  1          1   both   
4                    28.0         196                  1          1   both   

  product_name  aisle_id  department_id  prices price_range_loc  \
0         Soda        77              7     9.0             NaN   
1         Soda        77              7     9.0             NaN   
2         Soda        77              7     9.0             NaN   
3         Soda        77              7     9.0             NaN   
4         Soda        77              7     9.0             NaN   

      busiest_day  
0  Regularly busy  
1  Regularly busy  
2  Regularly busy  
3      Least busy  
4      Least busy

### Exercise 4.7

In [24]:
# Creating a for-loop for busiest days.

result2 = []

for value in df_ords_prods_merge['orders_day_of_week']:
    if value == 0 or value == 1:
        result2.append('Busiest days')
    elif value == 3 or value == 4:
        result2.append('Least busy days')
    else:
        result2.append('Regularly busy')

In [25]:
df_ords_prods_merge['busiest_days'] = result2  # Creating new column for "busiest_days".

In [26]:
df_ords_prods_merge['busiest_days'].value_counts(dropna = False)

busiest_days
Regularly busy     12916111
Busiest days       11864412
Least busy days     7624336
Name: count, dtype: int64

In [27]:
df_ords_prods_merge['orders_day_of_week'].value_counts(dropna = False) 

orders_day_of_week
0    6204182
1    5660230
6    4496490
2    4213830
5    4205791
3    3840534
4    3783802
Name: count, dtype: int64

#adding Saturday and Sunday (the two busiest days) gives us 11,864,412; addind Wednesday and Tuesday (the two slowest days) gives us 7,624,336 which are the least busiest days.

In [28]:
df_ords_prods_merge['order_hour_of_day'].value_counts(dropna = False) 

order_hour_of_day
10    2761760
11    2736140
14    2689136
15    2662144
13    2660954
12    2618532
16    2535202
9     2454203
17    2087654
8     1718118
18    1636502
19    1258305
20     976156
7      891054
21     795637
22     634225
23     402316
6      290493
0      218769
1      115700
5       87961
2       69375
4       53242
3       51281
Name: count, dtype: int64

In [29]:
# Creating a for-loop for busiest hours of day.

result3 = []

for value in df_ords_prods_merge['order_hour_of_day']:
    if value == 10 or value == 11 or value == 14 or value == 15 or value == 13 or value == 12 or value == 16 or value == 9:
        result3.append('Most orders')
    elif value == 17 or value == 8 or value == 18 or value == 19 or value == 20 or value == 7 or value == 21 or value == 22:
        result3.append('Average orders')
    else:
        result3.append('Fewest orders')

In [30]:
df_ords_prods_merge['busiest_period_of_day'] = result3  # Creating new column for 'busiest_period_of_day'

In [31]:
df_ords_prods_merge['busiest_period_of_day'].value_counts(dropna = False)

busiest_period_of_day
Most orders       21118071
Average orders     9997651
Fewest orders      1289137
Name: count, dtype: int64

In [32]:
df_ords_prods_merge.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'ords_prods_merge_updated.pkl'))